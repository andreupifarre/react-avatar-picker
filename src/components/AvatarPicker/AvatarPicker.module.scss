.AvatarPicker {
  margin: 0 auto;
  width: 60px;
  height: 60px;
  border-radius: 50%;
  user-select: none;
}

.AvatarImage {
  border: 1px solid transparent;
  width: inherit;
  height: inherit;
  border-radius: inherit;
}

.AvatarImageHovered,
.AvatarImage:hover,
.AvatarImage:focus {
  border-color: rgb(155, 160, 163);
}

.Popover {
  pointer-events: none;
  position: absolute;
  opacity: 0;
  transform-origin: 20px;
  transform: scale(0.8);
  transition: all 0.2s cubic-bezier(0.685, 1.555, 0.76, 1.545);
  transition-delay: 100ms;
  width: 0;
}

.PopoverExpanded {
  pointer-events: all;
  opacity: 1;
  // transition: all 0.15s ease-out;
  transform: scale(1);
  // transform-origin: 20px 5px;
}
