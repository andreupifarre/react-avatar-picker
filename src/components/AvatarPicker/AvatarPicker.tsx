import React, { useState, useRef } from 'react';
import classNames from 'classnames';
import Popover from '../Popover/Popover';
import AvatarImage from '../AvatarImage/AvatarImage';
import { AvatarImageItemType } from '../AvatarImage/AvatarImageItemType';
import useOutsideClick from '../../hooks/useOutsideClick';
import styles from './AvatarPicker.module.scss';

const defaultItem: number = 0;
const items: AvatarImageItemType[] = [
  { src: 'avatar1.png', label: 'Avatar 1', id: 1 },
  { src: 'avatar2.png', label: 'Avatar 2', id: 2 },
  { src: 'avatar3.png', label: 'Avatar 3', id: 3 },
  { src: 'avatar4.png', label: 'Avatar 4', id: 4 },
  { src: 'avatar5.png', label: 'Avatar 5', id: 5 },
  { src: 'avatar6.png', label: 'Avatar 6', id: 6 },
];

const AvatarPicker = () => {
  const ref = useRef<HTMLInputElement | null>(null);
  const [selected, setSelected] = useState(items[defaultItem]);
  const [isExpanded, setIsExpanded] = useState(false);
  const [isHovered, setIsHovered] = useState(false);
  const [isLoading, setIsLoading] = useState(-1);

  const handleClick = () => {
    setIsExpanded(!isExpanded);
    setIsHovered(!isHovered);
  };

  const handleSelected = (id: number) => {
    setIsLoading(id);
    setIsHovered(false);

    setTimeout(() => {
      setSelected(findItem(id));
      setIsExpanded(false);
      setIsLoading(-1);
    }, 1000);
  };

  const findItem = (id: number): AvatarImageItemType =>
    items.find((item: AvatarImageItemType) => item.id === id)!;

  useOutsideClick(ref, () => {
    setIsExpanded(false);
    setIsHovered(false);
  });

  return (
    <div ref={ref} className={styles.AvatarPicker}>
      <div
        className={classNames(styles.AvatarImage, {
          [styles.AvatarImageHovered]: isHovered,
        })}
      >
        <AvatarImage
          id={selected.id}
          src={selected.src}
          label={selected.label}
          onClick={handleClick}
        />
      </div>
      <div
        className={classNames(styles.Popover, {
          [styles.PopoverExpanded]: isExpanded,
        })}
      >
        <Popover
          items={items}
          onSelected={handleSelected}
          selected={selected.id}
          loading={isLoading}
        />
      </div>
    </div>
  );
};

export default AvatarPicker;
