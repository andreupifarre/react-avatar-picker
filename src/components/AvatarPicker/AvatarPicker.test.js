import React from 'react';
import { shallow } from 'enzyme';
import AvatarPicker from './AvatarPicker';

const wrapper = shallow(<AvatarPicker />);

describe('AvatarPicker', () => {
  it('should contain a AvatarImage component', () => {
    expect(wrapper.find('AvatarImage').exists()).toBeTruthy();
  });

  it('should contain a Popover component', () => {
    expect(wrapper.find('Popover').exists()).toBeTruthy();
  });

  it('should render correctly', () => {
    expect(wrapper.getElements()).toMatchSnapshot();
  });
});
