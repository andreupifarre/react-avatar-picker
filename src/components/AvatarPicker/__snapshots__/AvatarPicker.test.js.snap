// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarPicker should render correctly 1`] = `
Array [
  <div
    className="AvatarPicker"
  >
    <div
      className="AvatarImage"
    >
      <AvatarImage
        id={1}
        label="Avatar 1"
        onClick={[Function]}
        src="avatar1.png"
      />
    </div>
    <div
      className="Popover"
    >
      <Popover
        items={
          Array [
            Object {
              "id": 1,
              "label": "Avatar 1",
              "src": "avatar1.png",
            },
            Object {
              "id": 2,
              "label": "Avatar 2",
              "src": "avatar2.png",
            },
            Object {
              "id": 3,
              "label": "Avatar 3",
              "src": "avatar3.png",
            },
            Object {
              "id": 4,
              "label": "Avatar 4",
              "src": "avatar4.png",
            },
            Object {
              "id": 5,
              "label": "Avatar 5",
              "src": "avatar5.png",
            },
            Object {
              "id": 6,
              "label": "Avatar 6",
              "src": "avatar6.png",
            },
          ]
        }
        loading={-1}
        onSelected={[Function]}
        selected={1}
      />
    </div>
  </div>,
]
`;
